### PR TITLE
feat: add api-server authorization config

### PR DESCRIPTION
**Pull request by frezbo**
Add support for configuring KubeAPIServer authorization config.

Fixes: #9791

### CODE REVIEW

File: pkg/machinery/config/types/v1alpha1/v1alpha1_types.go [hunk 1]
```diff
@@ -1126,6 +1126,12 @@ type APIServerConfig struct {
 	//   schema:
 	//     type: object
 	ResourcesConfig *ResourcesConfig `yaml:"resources,omitempty"`
+	//   description: |
+	//     Configure the API server authorization config.
+	//     Node and RBAC authorizers are always added irrespective of the configuration.
```
Comment (smira) [inline pkg/machinery/config/types/v1alpha1/v1alpha1_types.go:1131]: I would join this with the first sentence, so it appears always in the docs.

File: internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go [hunk 2]
```diff
@@ -429,6 +429,29 @@ func (ctrl *ControlPlaneStaticPodController) manageAPIServer(ctx context.Context
 		"service-account-signing-key-file": argsbuilder.MergeDenied,
 		"tls-cert-file":                    argsbuilder.MergeDenied,
 		"tls-private-key-file":             argsbuilder.MergeDenied,
+		"authorization-config":             argsbuilder.MergeDenied,
+	}
+
+	k8sVersion := compatibility.VersionFromImageRef(cfg.Image)
+
+	if k8sVersion.KubeAPIServerSupportsAuthorizationConfigFile() {
+		if _, ok := cfg.ExtraArgs["authorization-mode"]; ok {
```
Comment (smira) [inline internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go:438]: I'm confused here a bit, if someone previously had `authorization-mode` set via extra args, I don't think we should stop supporting it, we should just keep it and (and same for `webhook-` args?)

File: internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go [hunk 3]
```diff
@@ -429,6 +429,29 @@ func (ctrl *ControlPlaneStaticPodController) manageAPIServer(ctx context.Context
...
+		}
+
+		for flag := range cfg.ExtraArgs {
+			if strings.HasPrefix(flag, "authorization-webhook-") {
+				return "", fmt.Errorf("%s flag is not supported when using authorization-config", flag)
+			}
+		}
+
+		if !k8sVersion.FeatureFlagStructuredAuthorizationConfigurationEnabledByDefault() {
+			builder.Set("feature-gates", "StructuredAuthorizationConfiguration=true")
```
Comment (smira) [inline internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go:449]: I guess we need to enable appending to the feature-gates? or is it fine?

File: pkg/machinery/resources/k8s/k8s.go [hunk 4]
```diff
@@ -7,7 +7,7 @@ package k8s
...
+//go:generate deep-copy -type AdmissionControlConfigSpec -type APIServerConfigSpec -type AuditPolicyConfigSpec -type BootstrapManifestsConfigSpec -type ConfigStatusSpec -type ControllerManagerConfigSpec -type EndpointSpec -type ExtraManifestsConfigSpec -type KubeletLifecycleSpec -type KubePrismConfigSpec -type KubePrismEndpointsSpec -type KubePrismStatusesSpec -type KubeletSpecSpec -type ManifestSpec -type ManifestStatusSpec -type NodeAnnotationSpecSpec -type NodeCordonedSpecSpec -type NodeLabelSpecSpec -type NodeTaintSpecSpec -type KubeletConfigSpec -type NodeIPSpec -type NodeIPConfigSpec -type NodeStatusSpec -type NodenameSpec -type SchedulerConfigSpec -type SecretsStatusSpec -type StaticPodSpec -type StaticPodStatusSpec -type StaticPodServerStatusSpec -type AuthorizationConfigSpec -header-file ../../../../hack/boilerplate.txt -o deep_copy.generated.go .
```
Comment (smira) [inline pkg/machinery/resources/k8s/k8s.go:10]: I tried to keep this sorted

File: internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go (hunk 3 above)
Comment (frezbo) [reply internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go:449]: it's already defined as `argsbuilder.MergeAdditive`, updated tests to reflect the same

File: internal/app/machined/pkg/controllers/k8s/render_config_static_pods.go [hunk 5]
```diff
@@ -99,6 +105,19 @@ func (ctrl *RenderConfigsStaticPodController) Run(ctx context.Context, r control
 
 		auditConfig := auditRes.TypedSpec()
 
+		authorizerConfigRes, err := safe.ReaderGetByID[*k8s.AuthorizationConfig](ctx, r, k8s.AuthorizationConfigID)
+		if err != nil {
+			if state.IsNotFoundError(err) {
```
[the remark this answers is not in the document]
Comment (frezbo) [reply internal/app/machined/pkg/controllers/k8s/render_config_static_pods.go:110]: i've a fix, so resource is always created, but not used for k8s versions not supporting it

File: internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go (hunk 2 above)
Comment (frezbo) [reply internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go:438]: i guess we can check if both are set and during machine config validation, I was wondering what I forgot to check

File: internal/app/machined/pkg/controllers/k8s/control_plane.go [hunk 6]
```diff
@@ -52,6 +54,28 @@ func controlplaneMapFunc[Output generic.ResourceWithRD](output Output) func(cfg
 	}
 }
 
+func controlPlaneMapFuncAuthorizationConfig() func(cfg *config.MachineConfig) optional.Optional[*k8s.AuthorizationConfig] {
```
[the remark this answers is not in the document]
Comment (frezbo) [reply internal/app/machined/pkg/controllers/k8s/control_plane.go]: fixed

File: internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go (hunk 2 above)
Comment (frezbo) [reply internal/app/machined/pkg/controllers/k8s/control_plane_static_pod.go]: handled all cases and added tests

File: go.mod [hunk 7]
```diff
@@ -148,7 +149,7 @@ require (
 	github.com/siderolabs/go-debug v0.4.0
 	github.com/siderolabs/go-kmsg v0.1.4
 	github.com/siderolabs/go-kubeconfig v0.1.0
-	github.com/siderolabs/go-kubernetes v0.2.16
+	github.com/siderolabs/go-kubernetes v0.2.17-0.20241127142321-06f07ab00042
```
Comment (smira) [inline go.mod:152]: it'd be nice to have a release (so we go into beta with everything released)

File: go.mod [hunk 8]
```diff
@@ -63,13 +63,14 @@ require (
...
 	github.com/containerd/errdefs v1.0.0
+	github.com/containerd/log v0.1.0
 	github.com/containerd/platforms v1.0.0-rc.0
 	github.com/containerd/typeurl/v2 v2.2.3
 	github.com/containernetworking/cni v1.2.3
 	github.com/containernetworking/plugins v1.6.0
 	github.com/coredns/coredns v1.11.3
 	github.com/coreos/go-iptables v0.8.0
-	github.com/cosi-project/runtime v0.7.1
+	github.com/cosi-project/runtime v0.7.2
```
Comment (smira) [inline go.mod:73]: let's sync machinery/go.mod as well

File: go.mod (hunk 7 above)
Comment (frezbo) [reply go.mod:152]: yup I can do that

File: pkg/machinery/config/types/v1alpha1/v1alpha1_types.go (hunk 1 above)
Comment (smira) [reply pkg/machinery/config/types/v1alpha1/v1alpha1_types.go:1131]: I think this is not fixed (still not visible, see e.g. `config.md`)

**Review by smira — APPROVED** [review]
:cool:

**Pull request merged**